File: client/src/js/app/spike/LoopbackProbe.ts
import { invoke } from '@tauri-apps/api/core';

/** Where the backend's throwaway listeners bound. */
interface ProbePorts {
    ws: number;
    http: number;
}

/** What the listening side observed, which the page cannot see for itself. */
interface ProbeStats {
    tcp_accepted: number;
    ws_upgraded: number;
    http_served: number;
}

/**
 * THROWAWAY. Whether this platform's webview may talk to a loopback socket in this app.
 *
 * Delete this file, `src-tauri/src/spike`, `commands/spike.rs` and the About pane card together.
 *
 * The whole design of the meter pivot rests on one unverified assumption: that a webview at the
 * app's production origin may open `ws://127.0.0.1`. Two platform rules could refuse it, and
 * they need different fixes — Android's cleartext policy, which a release build turns on and a
 * debug build turns off, and iOS App Transport Security, which applies to WebSocket since
 * WKWebView moved to the NSURLSession implementation. A refusal by either arrives in the page as
 * an untyped `error` event and close code 1006, so the page alone cannot name the cause.
 *
 * So the probe reads both sides. The backend counts sockets it accepted; the page counts frames
 * it received. A connection the page never opened and the backend never saw was stopped by a
 * platform policy before it left the process. One the backend accepted and the page still failed
 * is a protocol or handshake problem. That distinction is the entire output.
 *
 * Three hosts, because they do not behave alike: WKWebView has been reported to reach
 * `127.0.0.1` where it fails on `localhost`, and `[::1]` is a separate stack again.
 */
export class LoopbackProbe {
    /** How long a connection attempt is given before it is called a failure. */
    private static readonly ATTEMPT_TIMEOUT_MS = 4000;

    /** Frames to wait for after the socket opens, before calling it good. */
    private static readonly WANT_FRAMES = 2;

    async run(): Promise<string> {
        const lines: string[] = [];
        lines.push('=== BVC loopback WebSocket probe (throwaway) ===');
        lines.push(`when          ${new Date().toISOString()}`);
        lines.push(`origin        ${location.origin}`);
        lines.push(`protocol      ${location.protocol}`);
        lines.push(`href          ${location.href}`);
        lines.push(`secureContext ${String(window.isSecureContext)}`);
        lines.push(`userAgent     ${navigator.userAgent}`);
        lines.push('');

        let ports: ProbePorts;
        try {
            ports = await invoke<ProbePorts>('spike_probe_start');
            lines.push(`listeners     ws=${ports.ws} http=${ports.http}  (bound on 127.0.0.1)`);
        } catch (e) {
            lines.push(`listeners     FAILED TO BIND: ${LoopbackProbe.describe(e)}`);
            lines.push('');
            lines.push('The app could not bind a loopback socket at all. Nothing else was tried.');
            return lines.join('\n');
        }
        lines.push('');

        lines.push(`fetch http://127.0.0.1:${ports.http}/`);
        lines.push(`  ${await this.probeFetch(`http://127.0.0.1:${ports.http}/`)}`);
        lines.push('');

        for (const host of ['127.0.0.1', 'localhost', '[::1]']) {
            const url = `ws://${host}:${ports.ws}/metrics?key=probe`;
            lines.push(`ws    ${url}`);
            lines.push(`  ${await this.probeSocket(url)}`);
        }
        lines.push('');

        lines.push('--- candidate 2: the app\'s own URI scheme (never reaches the network stack) ---');
        for (const url of [
            'http://bvcpush.localhost/probe',
            'bvcpush://localhost/probe',
        ]) {
            lines.push(`push  ${url}`);
            lines.push(`  ${await this.probeFetch(url)}`);
        }
        try {
            const served = await invoke<number>('spike_push_served');
            lines.push(`  backend answered ${served} scheme request(s)`);
        } catch (e) {
            lines.push(`  backend count unreadable: ${LoopbackProbe.describe(e)}`);
        }
        lines.push('');

        try {
            const stats = await invoke<ProbeStats>('spike_probe_stats');
            lines.push(
                `backend saw   tcp_accepted=${stats.tcp_accepted} ws_upgraded=${stats.ws_upgraded} http_served=${stats.http_served}`,
            );
            lines.push('');
            lines.push(
                stats.tcp_accepted === 0
                    ? 'VERDICT ws    nothing reached the listener — a platform policy refused the connection.'
                    : stats.ws_upgraded === 0
                      ? 'VERDICT ws    sockets arrived but no WebSocket handshake completed — a protocol problem, not a policy one.'
                      : 'VERDICT ws    loopback WebSocket works at this origin.',
            );
        } catch (e) {
            lines.push(`backend saw   could not read: ${LoopbackProbe.describe(e)}`);
        }

        return lines.join('\n');
    }

    /**
     * A plain cleartext request to the same loopback address.
     *
     * Cleartext policy refuses this and the WebSocket alike; a WebSocket-specific rule refuses
     * only the socket. Which of the two is refused is the difference between a manifest change
     * and a redesign.
     */
    private async probeFetch(url: string): Promise<string> {
        const started = performance.now();
        try {
            const response = await fetch(url, { cache: 'no-store' });
            const body = await response.text();
            return `ok  status=${response.status} body=${JSON.stringify(body)} in ${LoopbackProbe.since(started)}`;
        } catch (e) {
            return `FAILED  ${LoopbackProbe.describe(e)}  after ${LoopbackProbe.since(started)}`;
        }
    }

    private probeSocket(url: string): Promise<string> {
        return new Promise<string>((resolve) => {
            const started = performance.now();
            let frames = 0;
            let opened = false;
            let settled = false;
            let socket: WebSocket;

            const finish = (verdict: string) => {
                if (settled) return;
                settled = true;
                clearTimeout(timer);
                try {
                    socket.close();
                } catch {
                    // A socket that never opened throws here on some engines; the verdict stands.
                }
                resolve(`${verdict}  after ${LoopbackProbe.since(started)}`);
            };

            const timer = setTimeout(
                () =>
                    finish(
                        opened
                            ? `OPENED but only ${frames} frames arrived  readyState=${socket.readyState}`
                            : `TIMED OUT before opening  readyState=${socket.readyState}`,
                    ),
                LoopbackProbe.ATTEMPT_TIMEOUT_MS,
            );

            try {
                socket = new WebSocket(url);
            } catch (e) {
                clearTimeout(timer);
                resolve(`THREW ON CONSTRUCT  ${LoopbackProbe.describe(e)}`);
                return;
            }

            socket.onopen = () => {
                opened = true;
            };
            socket.onmessage = (event) => {
                frames += 1;
                if (frames >= LoopbackProbe.WANT_FRAMES) {
                    finish(`ok  opened and received ${frames} frames, last=${String(event.data)}`);
                }
            };
            // Engines withhold the reason deliberately, so this records that it happened and
            // leaves the cause to the accept counters.
            socket.onerror = () =>
                finish(
                    opened
                        ? `ERROR after opening  ${frames} frames received`
                        : 'ERROR before opening  (no reason is exposed to script)',
                );
            socket.onclose = (event) =>
                finish(
                    `CLOSED  code=${event.code} reason=${JSON.stringify(event.reason)} clean=${String(event.wasClean)} opened=${String(opened)} frames=${frames}`,
                );
        });
    }

    private static since(started: number): string {
        return `${Math.round(performance.now() - started)}ms`;
    }

    private static describe(e: unknown): string {
        if (e instanceof Error) return `${e.name}: ${e.message}`;
        return String(e);
    }
}

File: client/src-tauri/src/spike/loopback_probe.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicU32, Ordering};

use serde::Serialize;
use tokio::io::AsyncWriteExt;
use tokio::net::TcpListener;

/// THROWAWAY. Measures whether this platform's webview may open a loopback socket at all.
///
/// Delete this module, `commands::spike`, `js/app/spike` and the About pane card with it once
/// the answer is recorded. Nothing else may depend on it.
///
/// The counters are the point. A webview refuses a connection with an untyped `error` event and
/// close code 1006 whatever the cause, so the page alone cannot separate a platform policy that
/// dropped the connection before it left the process from a handshake that failed after it
/// arrived. Counting accepts on this side answers that without a console, which iPadOS does not
/// readily give up.
#[derive(Default)]
pub struct LoopbackProbe {
    tcp_accepted: AtomicU32,
    ws_upgraded: AtomicU32,
    http_served: AtomicU32,
    ports: std::sync::Mutex<Option<ProbePorts>>,
}

/// Where the page should aim. Two listeners, because the interesting failure is asymmetric:
/// cleartext policy refuses both, and a WebSocket-specific rule refuses only the first.
#[derive(Debug, Clone, Copy, Serialize)]
pub struct ProbePorts {
    pub ws: u16,
    pub http: u16,
}

/// What the two sides each observed, for the page to render side by side.
#[derive(Debug, Clone, Copy, Serialize)]
pub struct ProbeStats {
    pub tcp_accepted: u32,
    pub ws_upgraded: u32,
    pub http_served: u32,
}

impl LoopbackProbe {
    /// How often a connected socket is sent a frame, and how many it is sent.
    ///
    /// Fast and finite: the page waits on the first few, and a probe listener must not outlive
    /// the panel that opened it.
    const PUSH_INTERVAL: std::time::Duration = std::time::Duration::from_millis(250);
    const PUSH_COUNT: u32 = 40;

    pub fn new_shared() -> Arc<Self> {
        Arc::new(Self::default())
    }

    /// Bind both listeners on loopback, or report why not.
    ///
    /// Idempotent: a second press reuses the ports rather than leaking a listener.
    pub async fn start(self: &Arc<Self>) -> Result<ProbePorts, anyhow::Error> {
        if let Some(ports) = self.ports.lock().ok().and_then(|slot| *slot) {
            return Ok(ports);
        }

        let ws_listener = TcpListener::bind("127.0.0.1:0").await?;
        let http_listener = TcpListener::bind("127.0.0.1:0").await?;
        let ports = ProbePorts {
            ws: ws_listener.local_addr()?.port(),
            http: http_listener.local_addr()?.port(),
        };

        self.spawn_ws_loop(ws_listener);
        self.spawn_http_loop(http_listener);

        if let Ok(mut slot) = self.ports.lock() {
            *slot = Some(ports);
        }
        log::info!(
            "loopback probe listening: ws=127.0.0.1:{} http=127.0.0.1:{}",
            ports.ws,
            ports.http
        );
        Ok(ports)
    }

    pub fn stats(&self) -> ProbeStats {
        ProbeStats {
            tcp_accepted: self.tcp_accepted.load(Ordering::Relaxed),
            ws_upgraded: self.ws_upgraded.load(Ordering::Relaxed),
            http_served: self.http_served.load(Ordering::Relaxed),
        }
    }

    fn spawn_ws_loop(self: &Arc<Self>, listener: TcpListener) {
        let probe = self.clone();
        tokio::spawn(async move {
            while let Ok((stream, peer)) = listener.accept().await {
                probe.tcp_accepted.fetch_add(1, Ordering::Relaxed);
                log::info!("loopback probe: ws socket accepted from {}", peer);
                let probe = probe.clone();
                tokio::spawn(async move {
                    if let Err(e) = probe.serve_ws(stream).await {
                        log::warn!("loopback probe: ws connection ended: {}", e);
                    }
                });
            }
        });
    }

    async fn serve_ws(&self, stream: tokio::net::TcpStream) -> Result<(), anyhow::Error> {
        use futures_util::SinkExt;

        let mut ws = tokio_tungstenite::accept_async(stream).await?;
        self.ws_upgraded.fetch_add(1, Ordering::Relaxed);
        log::info!("loopback probe: ws handshake completed");

        for seq in 0..Self::PUSH_COUNT {
            tokio::time::sleep(Self::PUSH_INTERVAL).await;
            let frame = format!("{{\"type\":\"probe\",\"seq\":{}}}", seq);
            ws.send(tokio_tungstenite::tungstenite::Message::Text(frame.into()))
                .await?;
        }

        ws.close(None).await?;
        Ok(())
    }

    /// A fixed 200, whatever is asked for. Its only job is to be reachable over cleartext, so a
    /// refused `fetch` and a refused WebSocket can be told apart.
    fn spawn_http_loop(self: &Arc<Self>, listener: TcpListener) {
        let probe = self.clone();
        tokio::spawn(async move {
            while let Ok((mut stream, peer)) = listener.accept().await {
                probe.tcp_accepted.fetch_add(1, Ordering::Relaxed);
                probe.http_served.fetch_add(1, Ordering::Relaxed);
                log::info!("loopback probe: http socket accepted from {}", peer);
                tokio::spawn(async move {
                    let body = "probe-ok";
                    let response = format!(
                        "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-Length: {}\r\nAccess-Control-Allow-Origin: *\r\nConnection: close\r\n\r\n{}",
                        body.len(),
                        body
                    );
                    let _ = stream.write_all(response.as_bytes()).await;
                    let _ = stream.shutdown().await;
                });
            }
        });
    }
}
